src/proxy.rs

```diff
@@ -298,36 +300,52 @@ pub fn get_original_src_from_stream(stream: &TcpStream) -> Option<IpAddr> {
         .map_or(None, |sa| Some(socket::to_canonical(sa).ip()))
 }
 
+const CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+
 pub async fn freebind_connect(local: Option<IpAddr>, addr: SocketAddr) -> io::Result<TcpStream> {
-    match local {
-        None => Ok(TcpStream::connect(addr).await?),
-        // TODO: Need figure out how to handle case of loadbalancing to itself.
-        //       We use ztunnel addr instead, otherwise app side will be confused.
-        Some(src) if src == socket::to_canonical(addr).ip() => Ok(TcpStream::connect(addr).await?),
-        Some(src) => {
-            let socket = if src.is_ipv4() {
-                TcpSocket::new_v4()?
-            } else {
-                TcpSocket::new_v6()?
-            };
-
-            let local_addr = SocketAddr::new(src, 0);
-            match socket::set_freebind_and_transparent(&socket) {
-                Err(err) => warn!("failed to set freebind: {:?}", err),
-                _ => {
-                    if let Err(err) = socket.bind(local_addr) {
-                        warn!("failed to bind local addr: {:?}", err)
+    async fn connect(local: Option<IpAddr>, addr: SocketAddr) -> io::Result<TcpStream> {
+        match local {
+            None => {
+                trace!(dest=%addr, "no local address, connect directly");
+                Ok(TcpStream::connect(addr).await?)
+            }
+            // TODO: Need figure out how to handle case of loadbalancing to itself.
+            //       We use ztunnel addr instead, otherwise app side will be confused.
+            Some(src) if src == socket::to_canonical(addr).ip() => {
+                trace!(%src, dest=%addr, "dest and source are the same, connect directly");
+                Ok(TcpStream::connect(addr).await?)
+            }
+            Some(src) => {
+                let socket = if src.is_ipv4() {
+                    TcpSocket::new_v4()?
+                } else {
+                    TcpSocket::new_v6()?
+                };
+
+                let local_addr = SocketAddr::new(src, 0);
+                match socket::set_freebind_and_transparent(&socket) {
+                    Err(err) => warn!("failed to set freebind: {:?}", err),
+                    _ => {
+                        if let Err(err) = socket.bind(local_addr) {
+                            warn!("failed to bind local addr: {:?}", err)
+                        }
                     }
-                }
-            };
-            Ok(socket.connect(addr).await?)
+                };
+                trace!(%src, dest=%addr, "connect with source IP");
+                Ok(socket.connect(addr).await?)
+            }
         }
     }
+    // Wrap the entire connect function in a timeout
+    timeout(CONNECTION_TIMEOUT, connect(local, addr))
```

Comment:
**question (non-blocking)**: should the timeout value be something configurable and it defaults to the constant if not configured?

**suggestion**: (if we do want this to be configurable) open an issue to track implementing a way to set the timeout